Integration of JSON type of GeoPackage

signedav opened the pull request:
## Description
GDAL 2.4 introduced "before" this Christmas supports the new JSON subtype. 

Integration is made with this PR. Reading and writing is possible with the widgets List (JSON arrays) and Key/Value (JSON Maps).
![listandmap](https://user-images.githubusercontent.com/28384354/45422344-3bd28b80-b690-11e8-88b1-952bc22111aa.png)

## Way of work
There is the new native type "JSON" described as "Map (JSON)" in `QgsOgrProvider`.

### Loading JSON field
On loading layer (by opening project or adding GPKG as Layer): In `QgsOgrProvider::loadFields()` it's checking the `OGR_Fld_GetSubType` if it's `OFTJSON` (if the OGR type is string). If so it stores for the field the "type" `QVariant::Map` and the "subtype" `QVariant::String`.

On opening the form: In `QgsOgrUtils::getOgrFeatureAttribute` it's checking if the "type" of the field is a `QVariant::Map` and converts from "json"-format to `QVariant`. In `QgsField::displayString<` it converts back from `QVariant` to `QJsonObject` and then to `QString`.

### Store JSON fields
On saving: In `QgsOgrProvider::changeAttributeValues` or `QgsOgrProvider::addFeaturePrivate` it checks if the GPKG field has "type" `OFTString` and "subtype" `OFSTJSON` and if so it conversts the field from `QMap` to `QVariant` and then from "json"-format to `QString` - if this has no success it tries to convert from `QList` to `QVariant` and to `QString` and in the end from `QString` to `QVariant` and with the "json"-decoding back to `QString`.

### Adding JSON attributes
On the layer properties on editing the attributes there is the possibility to choose "QMap (JSON)", and if so, on `QgsOgrProvider::addAttributeOGRLevel` if the attribute has type `QMap` the GPKG type is set to `OFTString` and the subtype to `OFSTJSON`.

In `QgsOgrUtils::readOgrFields()` it's checking for `OGR_Fld_GetSubType` if it's `OFTJSON` (if the OGR type is string). If so it stores for the field 


## Checklist
- [x] Reading from JSON fields
- [x] Writing to JSON fields
- [x] Adding and removing of JSON attributes on layer
- [x] GDAL version check
- [x] Testing of JSON reading / writing / adding

signedav committed:
reading ogr JSON basics

signedav committed:
basic tests for json type in gpkg

signedav committed:
basic writing json gpkg

signedav committed:
basic tests for json writing

signedav committed:
adding and removing json attributes

signedav committed:
check for GDAL Version and exclude by ifdef

signedav committed:
store plain if not valid

signedav committed:
fix typo

nyalldawson reviewed (commented):
Nice!

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -470,7 +476,8 @@ QgsOgrProvider::QgsOgrProvider( QString const &uri, const ProviderOptions &optio
       << QgsVectorDataProvider::NativeType( tr( "Decimal number (real)" ), QStringLiteral( "double" ), QVariant::Double, 0, nMaxDoubleLen, 0, nMaxDoublePrec )
-      << QgsVectorDataProvider::NativeType( tr( "Text (string)" ), QStringLiteral( "string" ), QVariant::String, 0, 65535 );
+      << QgsVectorDataProvider::NativeType( tr( "Text (string)" ), QStringLiteral( "string" ), QVariant::String, 0, 65535 )
+      << QgsVectorDataProvider::NativeType( tr( "Map (JSON)" ), QStringLiteral( "JSON" ), QVariant::Map, 0, 0, 0, 0, QVariant::String );
```
Should this be "JSON (string)"? JSON also supports arrays and the internal representation (in gpkg) is a string with a type hint.

signedav replied:
You mean just the label? Still working with `QVariant::Map` right?

m-kuhn replied:
I was talking about the label, @nyalldawson's comment is also valid (in the future we could think about another way of hinting type promotion to qgsogrprovider, saved in the qgis project, but that's a different story).

signedav replied:
To make it coherent to the Postgres types json and jsonb (at the moment "Map (json)" and "Map (jsonb)") I would suggest to use the following labels:
- Postgres json: "JSON (json)"
- Postgres jsonb: "JSON (jsonb)"
- GPKG string (json): "JSON (string)"
Because in Postgres it can be a list as well. And in future even value relation.
What do you think?

m-kuhn replied:
:+1: I like

m-kuhn commented on src/gui/editorwidgets/qgslistwidgetfactory.cpp:
```diff
@@ -44,5 +44,5 @@ QgsEditorConfigWidget *QgsListWidgetFactory::configWidget( QgsVectorLayer *vl, i
 {
   const QgsField field = vl->fields().field( fieldIdx );
-  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || field.typeName() == QLatin1String( "json" ) || field.typeName() == QLatin1String( "jsonb" ) ) && field.subType() != QVariant::Invalid ? 20 : 0;
+  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || field.typeName() == QLatin1String( "json" ) || field.typeName() == QLatin1String( "jsonb" ) || field.typeName() == QLatin1String( "JSON" ) ) && field.subType() != QVariant::Invalid ? 20 : 0;
```
Should this work on `field.type() == QVariant::Map /* and maybe || field.type() == QVariant::List */` instead? That would make it more generic for any NoSQL-style type.

signedav replied:
Yes. I'll do thta. 
But it's a fallback anyway because it first picks the keyvaluewidgettype...

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1009,6 +1016,7 @@ void QgsOgrProvider::loadFields()
     OGRFieldSubType ogrSubType = OFSTNone;
 
     QVariant::Type varType;
+    QVariant::Type varSubType = QVariant::Invalid;
```
Is this used anywhere?

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1509,13 +1531,32 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
+#if GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,4,0)
+          if ( OGR_Fld_GetSubType( fldDef ) == OFSTJSON )
+          {
+            stringValue = QString::fromUtf8( QJsonDocument::fromVariant( attrVal.toMap() ).toJson().data() );
```
How about

```cpp
switch ( attrVal.type() )
{
  case QVariant::Map:
    ...
  case QVariant::List:
  case QVariant::StringList:
    ...

  case QVariant::String: // or default:?
    ...
}
```

That would remove the requirement for string comparison.

signedav committed:
beautyfications and clean up

signedav committed:
improve value conversion to json-string

signedav committed:
better labeling of json native type (postgres and gpkg)

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1385,6 +1408,19 @@ OGRGeometryH QgsOgrProvider::ConvertGeometryIfNecessary( OGRGeometryH hGeom )
+  switch ( type )
+  {
+    case QVariant::Map:
+      return QString::fromUtf8( QJsonDocument::fromVariant( value.toMap() ).toJson().data() );
```
I think the following change will work for map and list and make the code simpler.

```suggestion
       return QString::fromUtf8( QJsonDocument::fromVariant( value ).toJson().constData() );
```

signedav replied:
I cannot confirm that. `QString::fromUtf8( QJsonDocument::fromVariant( value ).toJson().constData() );` returns me an empty string if the value has a map structure.

m-kuhn replied:
That is surprising.
The method http://doc.qt.io/qt-5/qjsondocument.html#fromVariant is not overloaded and just takes a single parameter QVariant (which should be one of QVariantMap, QVariantHash, QVariantList or QStringList). Can you see an explanation for the behavior you observed (I'm a bit lost)?

signedav replied:
Hm... Regarding the documentation I confirm your statement. And after some more tests it apparently works. I don't know what I did wrong 5 days ago, that I noticed the behavior that it does not work.

m-kuhn commented on src/providers/ogr/qgsogrprovider.cpp:
```diff
@@ -1385,6 +1408,19 @@ OGRGeometryH QgsOgrProvider::ConvertGeometryIfNecessary( OGRGeometryH hGeom )
   return OGR_G_ForceTo( hGeom, layerGeomType, nullptr );
 }
 
+QString QgsOgrProvider::jsonStringValue( const QVariant &value, const QVariant::Type &type )
```
Method should be const?

signedav replied:
Agree.

m-kuhn commented on src/core/qgsogrutils.cpp:
```diff
@@ -236,6 +244,16 @@ QVariant QgsOgrUtils::getOgrFeatureAttribute( OGRFeatureH ogrFet, const QgsField
+        //it has to be JSON
+        //it's null if no json format
+        if ( encoding )
+          value = QJsonDocument::fromJson( encoding->toUnicode( OGR_F_GetFieldAsString( ogrFet, attIndex ) ).toUtf8() ).toVariant();
```
I think the extra UTF8 conversion is not required

```suggestion
           value = QJsonDocument::fromJson( encoding->toUnicode( OGR_F_GetFieldAsString( ogrFet, attIndex ) ) ).toVariant();
```

signedav replied:
I did that to have a proper `QByteArray` in the end for `fromJson`.

m-kuhn replied:
Good point. Let's stick to that.

m-kuhn commented on src/core/qgsogrutils.cpp:
```diff
@@ -236,6 +244,16 @@ QVariant QgsOgrUtils::getOgrFeatureAttribute( OGRFeatureH ogrFet, const QgsField
+        if ( encoding )
+          value = QJsonDocument::fromJson( encoding->toUnicode( OGR_F_GetFieldAsString( ogrFet, attIndex ) ).toUtf8() ).toVariant();
+        else
+          value = QJsonDocument::fromJson( QString::fromUtf8( OGR_F_GetFieldAsString( ogrFet, attIndex ) ).toUtf8() ).toVariant();
```
Same here

```suggestion
          value = QJsonDocument::fromJson( QString::fromUtf8( OGR_F_GetFieldAsString( ogrFet, attIndex ) ) ).toVariant();
```

m-kuhn commented on src/gui/editorwidgets/qgslistwidgetfactory.cpp:
```diff
@@ -44,5 +44,5 @@ QgsEditorConfigWidget *QgsListWidgetFactory::configWidget( QgsVectorLayer *vl, i
 {
   const QgsField field = vl->fields().field( fieldIdx );
-  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || field.typeName() == QLatin1String( "json" ) || field.typeName() == QLatin1String( "jsonb" ) ) && field.subType() != QVariant::Invalid ? 20 : 0;
+  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || QVariant::Map ) && field.subType() != QVariant::Invalid ? 20 : 0;
```
Whooops, a boolean check on a constant will always be the same (true here) `if (QVariant::Map)` will always be true.
```suggestion
  return ( field.type() == QVariant::List || field.type() == QVariant::StringList || field.type() == QVariant::Map ) && field.subType() != QVariant::Invalid ? 20 : 0;
```

signedav replied:
oh oops. of course...

signedav committed:
fix in listwidgetfactory and set method of ogrprovider to const not changing class values

signedav committed:
use general json conversion because it works for the different types

like QVariantMap, QVariantHash, QVariantList or QStringList
as fallback still the QString conversion is available
